stdlib/public/core/Random.swift

```diff
@@ -0,0 +1,167 @@
+/// the thread safety and quality of the generator.
+public protocol RandomNumberGenerator {
+  /// Returns a value from a uniform, independent distribution of binary data.
+  ///
+  /// - Returns: An unsigned 64-bit random value.
+  mutating func next() -> UInt64
+}
+
+extension RandomNumberGenerator {
+  /// Returns a value from a uniform, independent distribution of binary data.
+  ///
+  /// - Returns: A random value of `T`. Bits are randomly distributed so that
+  ///   every value of `T` is equally likely to be returned.
+  @inlinable
+  public mutating func next<T: FixedWidthInteger & UnsignedInteger>() -> T {
+    if T.bitWidth <= UInt64.bitWidth {
+      return T(truncatingIfNeeded: self.next())
+    }
+
+    let (quotient, remainder) = T.bitWidth.quotientAndRemainder(
+      dividingBy: UInt64.bitWidth
+    )
+    var tmp: T = 0
+
+    for i in 0 ..< quotient {
+      tmp += T(truncatingIfNeeded: next()) &<< (UInt64.bitWidth * i)
+    }
+
+    if remainder != 0 {
+      let random = next()
+      let mask = UInt64.max &>> (UInt64.bitWidth - remainder)
+      tmp += T(truncatingIfNeeded: random & mask) &<< (UInt64.bitWidth * quotient)
+    }
+
+    return tmp
+  }
+
+  /// Returns a random value that is less than the given upper bound.
+  ///
+  /// - Parameter upperBound: The upper bound for the randomly generated value.
+  /// - Returns: A random value of `T` in the range `0..<upperBound`. Every
+  ///   value in the range `0..<upperBound` is equally likely to be returned.
+  @inlinable
+  public mutating func next<T: FixedWidthInteger & UnsignedInteger>(
+    upperBound: T
+  ) -> T {
+    guard upperBound != 0 else { return 0 }
+    let tmp = (T.max % upperBound) + 1
+    let range = tmp == upperBound ? 0 : tmp
+    var random: T = 0
+
+    repeat {
+      random = next()
+    } while random < range
+
+    return random % upperBound
+  }
+}
+
+/// The default source of random data.
+///
+/// When you generate random values, shuffle a collection, or perform another
+/// operation that depends on random data, this type's `default` property is
+/// the generator used by default. For example, the two method calls in this
+/// example are equivalent:
+///
+///     let x = Int.random(in: 1...100)
+///     let y = Int.random(in: 1...100, using: &Random.default)
+///
+/// `Random.default` is safe to use in multiple threads, and uses a
+/// cryptographically secure algorithm whenever possible.
+///
+/// Platform Implementation of `Random`
+/// ===================================
+///
+/// - Apple platforms use `arc4random_buf(3)` for newer versions of their OS,
+///   and for older versions they will read from `/dev/urandom`.
+/// - `Linux`, `Android`, `Cygwin`, `Haiku`, `FreeBSD`, and `PS4` all try to
+///   use `getrandom(2)`, but if it doesn't exist then they read from
+///   `/dev/urandom`.
+/// - `Fuchsia` calls `getentropy(3)`.
+/// - `Windows` calls `BCryptGenRandom`.
+public struct Random : RandomNumberGenerator {
```

Comment:
I wonder how much resilience costs us for `Random`; `Random.default.next()` involves an indirectly-sized temporary and three resilient function calls, when we only really need a single call.

Do we expect that `Random` may become non-empty in the future? If not, we should make it `@_fixed_layout`, and we should make the initializer and `.default`'s getter and setter inlinable.

Reply:
Is there a way to mark a private initializer as `@inlinable` or something similar? I can't use a private initializer in an `@inlinable` getter/setter for a var, but I can if I make the initializer internal with `@usableFromInline`. I wonder if we have to make this change in order to achieve `@inlinable` for the static var.

Reply:
If we're okay with making it inlinable, I think `@inlinable internal` is the way to do it.